### PR TITLE
Disallow mixing nullish coalescing with logical OR via early errors

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
@codehag @rkirsling @jridgewell

### CODE REVIEW

File: spec.html [hunk 1]
```diff
@@ -29,6 +29,24 @@ <h1>Syntax</h1>
   </emu-grammar>
 </emu-clause>
 
+<emu-clause id="sec-binary-logical-operators-static-semantics-early-errors">
+  <h1>Static Semantics: Early Errors</h1>
+
+  <emu-grammar>LogicalORExpression : LogicalORExpression `||` LogicalANDExpression</emu-grammar>
+  <ul>
+    <li>
+      It is a Syntax Error if the derived |LogicalORExpression| on is <emu-grammar>LogicalORExpression : LogicalORExpression `??` LogicalANDExpression</emu-grammar>.
```
Comment (rkirsling) [inline spec.html]:
```suggestion
      It is a Syntax Error if the derived |LogicalORExpression| is <emu-grammar>LogicalORExpression : LogicalORExpression `??` LogicalANDExpression</emu-grammar>.
```

File: spec.html [hunk 2]
```diff
@@ -29,6 +29,24 @@ <h1>Syntax</h1>
...
+    </li>
+  </ul>
+
+  <emu-grammar>LogicalORExpression : LogicalORExpression `??` LogicalANDExpression</emu-grammar>
+  <ul>
+    <li>
+      It is a Syntax Error if the derived |LogicalORExpression| on is <emu-grammar>LogicalORExpression : LogicalORExpression `||` LogicalANDExpression</emu-grammar>.
```
Comment (rkirsling) [inline spec.html]:
```suggestion
      It is a Syntax Error if the derived |LogicalORExpression| is <emu-grammar>LogicalORExpression : LogicalORExpression `||` LogicalANDExpression</emu-grammar>.
```

**Comment by ljharb** [comment]
Are there issues this closes or relates to?

**Comment by rkirsling** [comment]
#26 and #15 are the relevant ones.

**Comment by ljharb** [comment]
Why not give them the same precedence, assuming https://github.com/tc39/proposal-nullish-coalescing/issues/15#issuecomment-373478940 isn't wrong?

**Comment by rkirsling** [comment]
Same precedence is the status quo, but it's problematic for the reasons laid out in https://github.com/tc39/proposal-nullish-coalescing/issues/26#issue-305360549.

Of the solutions discussed there, my personal vote would be to lower `??`'s precedence, but I don't object to this approach either.

**Comment by claudepache** [comment]
The solution proposed in this PR is wrong, because it is treating the symptom (mixing `??` and `||` does the wrong thing), and not the root of the disease (the precedence of `??` is poorly chosen).

One of the problems provoked by the wrong precedence and not solved by this PR is apparent when comparing  `!(maybeA ?? b && c)` and `(maybeNotA ?? notB || notC)`.

**Comment by DanielRosenwasser** [comment]
From the last TC39 meeting, we had some general agreement that disallowing the mix of both would be acceptable and would permit us to revisit at a later time. So the rationale to disallow the mix of both is:

1. We did the same for exponentiation and unary `+`/`-`.
2. It leaves room for design open if someone cares to do so later on.

My preference would probably be to align with other lower-precedence languages because same precedence feels like it would rarely be what I would want. I also get the sense that it's the behavior I've heard most others (both inside and outside of the committee) favor.

If there are no hard objections to lower precedence, I'm willing to bring that up at the next meeting and survey the room. However, I think that forbidding mixing the two is probably the best way to move forward and make progress.

**Comment by DanielRosenwasser** [comment]
Any objections to the contents of the spec text? @codehag and @rkirsling?

**Comment by jridgewell** [comment]
I think we should ban mixing `??` and `&&` as well. It's not enough that `&&` has always been tighter than `||` (and so of course it's tighter than `??`), because people still get that precedence incorrect.

Kevin Gibbons created a Grammar Flag parsing at the last meeting that would work perfectly: https://gist.github.com/bakkot/332ddeb26978c41613abc092120448b9

**Comment by ljharb** [comment]
@jridgewell would that mean that parens are always syntactically required to mix ?? with ||/&&?

**Comment by jridgewell** [comment]
Yes

**Comment by DanielRosenwasser** [comment]
I can add the nullish grammar flag, but I wonder if it's not just cleaner to issue the early error.

**Comment by DanielRosenwasser** [comment]
#39 changes the grammar directly, but I'm not sure 4 grammar parameters is the prettiest thing I've seen in this spec.

**Comment by rkirsling** [comment]
I believe the spec text here says what it intends to; it's merely a question of what we want the state of the proposal to be during the upcoming meeting:

- I think this is a reasonable state to have things be in if we're aiming to "pick up from where we left off", as the discussion last time seemed to center around `??` being `||`-like.

- One could also argue for leaving things as they are for now, since many of us are hoping to lower precedence instead, in which case the change here would be unnecessary. 

- I don't feel comfortable making a change that affects `&&` so hastily—`??` is not `&&`-like, so this ought to require a discussion of its own. (Although I guess it's taken me long enough to write this that #39 already exists, so I can express my thoughts over there.)

Overall though, it would be nice to allow  @DanielRosenwasser to try for stage 3 with the status quo reflecting _some_ solution to #26. :joy:

**Comment by bakkot** [comment]
@claudepache 

> One of the problems provoked by the wrong precedence and not solved by this PR is apparent when comparing `!(maybeA ?? b && c)` and `(maybeNotA ?? notB || notC)`.

I don't understand the problem this is pointing at; can you expand? The thing that I get when comparing those under this PR is "the first one behaves reasonably" and "the second one is not valid JS", which, while it is perhaps slightly surprising while performing that transformation, doesn't seem terrible.

**Comment by claudepache** [comment]
@bakkot Consider some expression mixing operators of different categories:

```js
A && B == C 
```

You have to learn whether it means `(A && B) == C` or `A && (B == C)`, but once you know that, you have strong expectations about the semantics of the following expression:

```js
A || B == C
```

It would be surprising if the latter triggers a syntax error.

But it does work as expected, because every comparison operators (`==`, `<=`, etc.) have a strictly higher precedence level than every logical operator (`||`, `&&`).

-----

Returning to the original issue. The root of the problem is that the following is currently *not* true: “Logical operators (`||`, `&&`) have either a strictly higher, or a strictly lower precedence level than nullish-coalescing.”

**Comment by bakkot** [comment]
@claudepache Thanks for expanding, that makes sense to me. Your last statement applies equally to #39, strictly speaking (which bans mixing with both `||` and `&&`); do you have the same issue with it?

**Comment by bakkot** [comment]
@DanielRosenwasser FWIW I find the grammar flags mich cleaner than prose Early Errors, but that's an editorial question which can be resolved at a later stage.

**Comment by claudepache** [comment]
> @bakkot Your last statement applies equally to #39, strictly speaking

@bakkot. My last statement is incomplete, because it overlooked the possibility of completely disallowing to mix of operators of two distinct categories (in which case, speaking of their relative precedence level is meaningless). — So, no, #39 has not the same issue.

**Comment by ljharb** [comment]
?? is in the same category as || imo.

**Pull request closed without merging**